src/pylint_odoo/checkers/function_checker.py

```diff
@@ -0,0 +1,46 @@
+from astroid import ClassDef, FunctionDef
+from pylint.checkers import utils
+
+from .odoo_base_checker import OdooBaseChecker
+
+ODOO_MSGS = {
+    # C->convention R->refactor W->warning E->error F->fatal
+    "E8501": (
+        "No exceptions should be raised inside unlink() functions",
+        "no-raise-unlink",
+        "Use @api.ondelete to add any constraints instead",
+    ),
+}
+
+
+class FunctionChecker(OdooBaseChecker):
+    msgs = ODOO_MSGS
+    name = "odoolint"
+
+    @staticmethod
+    def _is_unlink(node: FunctionDef) -> bool:
+        parent = getattr(node, "parent", False)
+        return (
+            isinstance(parent, ClassDef)
+            and ("_name" in parent.locals or "_inherit" in parent.locals)
+            and node.name == "unlink"
+        )
+
+    @staticmethod
+    def get_enclosing_function(node, depth=10):
+        count = 0
+        parent = getattr(node, "parent", False)
+        while not isinstance(parent, FunctionDef) and count < depth:
```

Comment:
I'd suggest using while's when there's no other alternative. In this case, this could be achieved by using

`range(depth)`

On the other hand, why is it required to recursively check parents?

Reply:
Parents are recursively checked to find the enclosing function, for example:

```python
def unlink(self):
    if self.business_logic():
        raise ValueError()  #  offending node, will be reviewed in visit_raise
```

In that case the initial node will be the `raise` but we need to know if it is inside an unlink function so the solution I came up with was to go up the parent tree (effectively up in terms of code lines) until a function definition is found (or the lookup limit is reached).